lib/transmuxer/ec3_transmuxer.js

```diff
@@ -0,0 +1,208 @@
+    }
+    return mimeType;
+  }
+
+
+  /**
+   * @override
+   * @export
+   */
+  getOrginalMimeType() {
+    return this.originalMimeType_;
+  }
+
+
+  /**
+   * @override
+   * @export
+   */
+  transmux(data, stream, reference, duration) {
+    const Ec3 = shaka.transmuxer.Ec3;
+    const Uint8ArrayUtils = shaka.util.Uint8ArrayUtils;
+
+    const uint8ArrayData = shaka.util.BufferUtils.toUint8(data);
+
+    const id3Data = shaka.util.Id3Utils.getID3Data(uint8ArrayData);
+    let offset = id3Data.length;
+    for (; offset < uint8ArrayData.length; offset++) {
+      if (Ec3.probe(uint8ArrayData, offset)) {
+        break;
+      }
+    }
+
+    let timestamp = reference.endTime * 1000;
+
+    const frames = shaka.util.Id3Utils.getID3Frames(id3Data);
+    if (frames.length && reference) {
+      const metadataTimestamp = frames.find((frame) => {
+        return frame.description ===
+            'com.apple.streaming.transportStreamTimestamp';
+      });
+      if (metadataTimestamp) {
+        timestamp = /** @type {!number} */(metadataTimestamp.data);
+      }
+    }
+
+    /** @type {number} */
+    let sampleRate = 0;
+
+    /** @type {!Uint8Array} */
+    let audioConfig = new Uint8Array([]);
+
+    /** @type {!Array.<shaka.util.Mp4Generator.Mp4Sample>} */
+    const samples = [];
+
+    while (offset < uint8ArrayData.length) {
+      const frame = Ec3.parseFrame(uint8ArrayData, offset);
+      if (!frame) {
+        return Promise.reject(new shaka.util.Error(
+            shaka.util.Error.Severity.CRITICAL,
+            shaka.util.Error.Category.MEDIA,
+            shaka.util.Error.Code.TRANSMUXING_FAILED));
+      }
+      stream.audioSamplingRate = frame.sampleRate;
+      stream.channelsCount = frame.channelCount;
+      sampleRate = frame.sampleRate;
+      audioConfig = frame.audioConfig;
+
+      const frameData = uint8ArrayData.subarray(
+          offset, offset + frame.frameLength);
+
+      samples.push({
+        data: frameData,
+        size: frame.frameLength,
+        duration: Ec3.EC3_SAMPLES_PER_FRAME,
+        cts: 0,
+        flags: {
+          isLeading: 0,
+          isDependedOn: 0,
+          hasRedundancy: 0,
+          degradPrio: 0,
+          dependsOn: 2,
+          isNonSync: 0,
+        },
+      });
+      offset += frame.frameLength;
+    }
+    /** @type {number} */
+    const baseMediaDecodeTime = Math.floor(timestamp * sampleRate / 1000);
+
+    /** @type {shaka.util.Mp4Generator.StreamInfo} */
+    const streamInfo = {
+      timescale: sampleRate,
+      duration: duration,
+      videoNalus: [],
+      audioConfig: audioConfig,
+      data: {
+        sequenceNumber: this.frameIndex_,
+        baseMediaDecodeTime: baseMediaDecodeTime,
+        samples: samples,
+      },
+      stream: stream,
+    };
+    const mp4Generator = new shaka.util.Mp4Generator(streamInfo);
+    let initSegment;
+    if (!this.initSegments.has(stream.id)) {
+      initSegment = mp4Generator.initSegment();
```

Comment:
If no stream-specific data is used when generating the init segment, is there any reason to make and store a separate init segment per stream? Couldn't we just generate a single init segment if there was none before, and use that for all streams?

Reply:
It cannot be done, the initialization segment depends on some parameters of the stream, for example, the audioConfig depends on values of the segment itself.

Reply:
Ah, I see, so it's based on the streamInfo passed to the Mp4Generator object. I had misread that as being a static method.